### PR TITLE
Switch to our fork of dh-virtualenv for compatibility with Python 3.11

### PR DESCRIPTION
**Pull request by anoadragon453**
The latest version of https://github.com/spotify/dh-virtualenv is incompatible with Python 3.11. We've [submitted a fix upstream](https://github.com/spotify/dh-virtualenv/pull/354), but in the meantime have put up a fork at https://github.com/matrix-org/dh-virtualenv/ with the fix on the `anoa/drop_getargspec` branch. See [this PR](https://github.com/matrix-org/dh-virtualenv/pull/1)'s description for a detailed description of the technical issues.

This PR switches Synapse over to that branch until the fix is merged and released upstream. It also updates our locked dependency of `cffi` from 1.15.0 to 1.15.1 to fix a mysterious build issue when building the debs on Debian sid.

### CODE REVIEW

**Comment by anoadragon453** [comment]
Oh, exciting! Well I wouldn't delay merging this if it becomes ready first. It'd be pretty easy to switch back to upstream even if they release today, but I'd like to unblock Synapse development ASAP.

**Comment by anoadragon453** [comment]
Hmm, well let's hope they do do a release, because the build has failed and I have no clue as to why.

**Comment by anoadragon453** [comment]
Ah, I tagged `master` + the fix, but we were previously running `1.2.2`, which is 40+ commits behind master.

Will do another tag.

**Comment by anoadragon453** [comment]
One can reproduce the build failure locally with:

```
docker build --tag dh-venv-builder:sid --build-arg distro=debian:sid -f docker/Dockerfile-dhvirtualenv docker
mkdir -p debs
docker run --rm --name 'synapse_build_sid' --volume="$PWD:/synapse/source:ro" --volume="$PWD/debs:/debs" -e 'TARGET_USERID=1001' -e 'TARGET_GROUPID=123' -e 'DEB_BUILD_OPTIONS=' dh-venv-builder:sid
```

Currently having a play with this local setup.

**Comment by anoadragon453** [comment]
So, for whatever reason `cffi==1.15.0` fails to build on Debian sid, whereas `cffi=1.15.1` works fine (might be a Python 3.11 incompatibility?). [The changelog](https://cffi.readthedocs.io/en/latest/whatsnew.html?highlight=changes) gives nothing away, but bumping the version fixes the mysterious build failure :tada:

**Comment by anoadragon453** [comment]
CI failures are (finally) unrelated to the proposed changes.

File: poetry.lock [hunk 1]
```diff
@@ -133,7 +134,7 @@ optional = false
 python-versions = ">=3.5.0"
 
 [package.extras]
-unicode-backport = ["unicodedata2"]
+unicode_backport = ["unicodedata2"]
```
Comment (clokep) [inline poetry.lock:137]: There's a lot of unrelated changes, IIRC this happens if you're not on the latest poetry? (Or maybe if you have weird things in your poetry cache?)

Can you try clearing your cache and recreating this something like:

```bash
$ poetry cache list
$ poetry cache clear _default_cache --all
$ poetry cache clear pypi --all
$ poetry lock --no-update
```

I'm a bit confused at how this file was modified though. Did you do `poetry update cffi==1.15.1` and then `poetry lock`?

Comment (anoadragon453) [reply poetry.lock:137]: Ah, yes. I ended up having to downgrade to poetry 1.2.0 to update the lock file, otherwise the docker image complained about the poetry lock file being too new. I don't know if we normally use poetry 1.3.0 to manipulate our lockfile though? And the solution is to update poetry in the docker image?

All I did here was run `poetry update cffi` on poetry 1.2.0.

Comment (clokep) [reply poetry.lock:137]: I think you want to use poetry `1.2.2` (one of the point releases changed something.)

**Review by clokep — APPROVED** [review]

File: poetry.lock (hunk 1 above)
Comment (clokep) [reply poetry.lock]: That sounds reasonable to me!
